atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/ReadValidationCommitTimestampLoader.java

```diff
@@ -0,0 +1,157 @@
+                delegate.getCommitTimestampsNonBlockingForValidation(tableRef, startTimestamps),
+                startToCommitTimestamps -> {
+                    if (startToCommitTimestamps.keySet().containsAll(startTimestamps)) {
+                        return startToCommitTimestamps;
+                    }
+                    // If we do not get back all these results we may be in the deadlock case so we
+                    // should just fail out early.  It may be the case that abort more transactions
+                    // than needed to break the deadlock cycle, but this should be pretty rare.
+                    transactionOutcomeMetrics.markReadWriteConflict(tableRef);
+                    throw new TransactionSerializableConflictException(
+                            "An uncommitted conflicting read was written after our start timestamp for table "
+                                    + tableRef + ".  This case can cause deadlock and is very likely to be a "
+                                    + "read write conflict.",
+                            tableRef);
+                },
+                MoreExecutors.directExecutor());
+    }
+
+    /**
+     * Partitions {@code startTimestamps} in two sets, based on their relation to the start timestamp provided.
+     *
+     * @param myStart start timestamp of this transaction
+     * @param startTimestamps of transactions we are interested in
+     * @return a {@link PartitionedTimestamps} object containing split timestamps
+     */
+    private PartitionedTimestamps splitTransactionBeforeAndAfter(long myStart, LongIterable startTimestamps) {
+        ImmutablePartitionedTimestamps.Builder builder =
+                ImmutablePartitionedTimestamps.builder().myCommitTimestamp(commitTs);
+        MutableLongSet beforeStart = LongSets.mutable.empty();
+        MutableLongSet afterStart = LongSets.mutable.empty();
+        startTimestamps.forEach(startTimestamp -> {
+            if (startTimestamp == myStart) {
+                builder.splittingStartTimestamp(myStart);
+            } else if (startTimestamp < myStart) {
+                beforeStart.add(startTimestamp);
+            } else {
+                afterStart.add(startTimestamp);
+            }
+        });
+        builder.beforeStart(beforeStart.asUnmodifiable());
+        builder.afterStart(afterStart.asUnmodifiable());
+        return builder.build();
+    }
+
+    @Override
+    public ListenableFuture<LongLongMap> getCommitTimestamps(
+            @Nullable TableReference tableRef, LongIterable startTimestamps) {
+        return getCommitTimestampsInternal(tableRef, startTimestamps, true);
+    }
+
+    @Override
+    public ListenableFuture<LongLongMap> getCommitTimestampsNonBlockingForValidation(
+            @Nullable TableReference tableRef, LongIterable startTimestamps) {
+        return getCommitTimestampsInternal(tableRef, startTimestamps, false);
+    }
```

Comment:
nit: I think we usually have the public methods coming first on the file by convention?

Reply:
my bad, fixed!